Detectors/CPV/calib/testWorkflow/GainCalibratorSpec.h

```diff
@@ -0,0 +1,190 @@
+// License v3 (GPL Version 3), copied verbatim in the file "COPYING".
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+#include "Framework/Task.h"
+#include "Framework/ConfigParamRegistry.h"
+#include "Framework/ControlService.h"
+#include "Framework/WorkflowSpec.h"
+#include "Framework/CCDBParamSpec.h"
+#include "CCDB/CcdbApi.h"
+#include "CCDB/CcdbObjectInfo.h"
+#include "DetectorsCalibration/Utils.h"
+#include "CPVCalibration/GainCalibrator.h"
+#include "DataFormatsCPV/Digit.h"
+#include "DataFormatsCPV/TriggerRecord.h"
+#include "DetectorsBase/GRPGeomHelper.h"
+
+using namespace o2::framework;
+
+namespace o2
+{
+namespace calibration
+{
+class CPVGainCalibratorSpec : public o2::framework::Task
+{
+ public:
+  CPVGainCalibratorSpec(std::shared_ptr<o2::base::GRPGeomRequest> req) : mCCDBRequest(req) {}
+
+  //_________________________________________________________________
+  void init(o2::framework::InitContext& ic) final
+  {
+    o2::base::GRPGeomHelper::instance().setRequest(mCCDBRequest);
+    // auto slotL = ic.options().get<uint32_t>("tf-per-slot");
+    // auto delay = ic.options().get<uint32_t>("max-delay");
+    auto updateInterval = ic.options().get<uint32_t>("updateInterval"); // in TF
+    bool updateAtTheEndOfRunOnly = ic.options().get<bool>("updateAtTheEndOfRunOnly");
+    mCalibrator = std::make_unique<o2::cpv::GainCalibrator>();
+    mCalibrator->setSlotLength(0);
+    mCalibrator->setMaxSlotsDelay(1000);
+    if (updateAtTheEndOfRunOnly) {
+      mCalibrator->setUpdateAtTheEndOfRunOnly();
+    }
+    mCalibrator->setCheckIntervalInfiniteSlot(updateInterval);
+    mCalibrator->setUpdateTFInterval(updateInterval);
+    LOG(info) << "CPVGainCalibratorSpec initialized";
+    LOG(info) << "tf-per-slot = 0 (inconfigurable for this calibrator)";
+    LOG(info) << "max-delay = 1000 (inconfigurable for this calibrator)";
+    LOG(info) << "updateInterval = " << updateInterval;
+    LOG(info) << "updateAtTheEndOfRunOnly = " << updateAtTheEndOfRunOnly;
+  }
+
+  void finaliseCCDB(o2::framework::ConcreteDataMatcher& matcher, void* obj) final
+  {
+    o2::base::GRPGeomHelper::instance().finaliseCCDB(matcher, obj);
+  }
+
+  //_________________________________________________________________
+  void run(o2::framework::ProcessingContext& pc) final
+  {
+    o2::base::GRPGeomHelper::instance().checkUpdates(pc);
+    o2::base::TFIDInfoHelper::fillTFIDInfo(pc, mCalibrator->getCurrentTFInfo());
+    TFType tfcounter = mCalibrator->getCurrentTFInfo().startTime;
+    auto&& digits = pc.inputs().get<gsl::span<o2::cpv::Digit>>("calibdigs");
+
+    // previous gains
+    if (!mCalibrator->isSettedPreviousGains()) {
+      const auto previousGains = o2::framework::DataRefUtils::as<CCDBSerialized<o2::cpv::CalibParams>>(pc.inputs().get("gains"));
```

Comment:
please always use `pc.inputs().get<o2::cpv::CalibParams*>("gains")` rather than `as<...>` which does not invoke finaliseCCDB callback and may miss other features in the future.